File: src/main/java/dev/isxander/evergreenhud/elements/type/SimpleTextElement.java
/*
 * Copyright (C) isXander [2019 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.elements.type;

import dev.isxander.evergreenhud.settings.impl.*;
import dev.isxander.xanderlib.utils.GuiUtils;
import dev.isxander.xanderlib.utils.Resolution;
import dev.isxander.xanderlib.utils.json.BetterJsonObject;
import lombok.Getter;
import net.minecraft.client.renderer.GlStateManager;

/**
 * Renders a single line of text to the screen
 */
public abstract class SimpleTextElement extends TextElement {

    @Getter private BooleanSetting invertedTitle;

    @Override
    public void render(float partialTicks, int origin) {
        super.render(partialTicks, origin);

        String displayString = getDisplayString();
        float scale = getPosition().getScale();
        boolean chroma = getChroma().get();
        TextMode textMode = getTextMode().get();
        int color = getTextColor().getRGB();

        float x = getPosition().getRawX(Resolution.get());
        float y = getPosition().getRawY(Resolution.get());
        GlStateManager.pushMatrix();
        GlStateManager.scale(scale, scale, 0);

        float posX = x / scale;
        float posY = y / scale;
        switch (getAlignment().get()) {
            case RIGHT:
                posX = (x - mc.fontRendererObj.getStringWidth(displayString)) / scale;
                GuiUtils.drawString(mc.fontRendererObj, displayString, posX, posY, textMode == TextMode.SHADOW, textMode == TextMode.BORDER, chroma, false, color);

                break;
            case CENTER:
                GuiUtils.drawString(mc.fontRendererObj, displayString, posX, posY, textMode == TextMode.SHADOW, textMode == TextMode.BORDER, chroma, true, color);

                break;
            case LEFT:
                GuiUtils.drawString(mc.fontRendererObj, displayString, posX, posY, textMode == TextMode.SHADOW, textMode == TextMode.BORDER, chroma, false, color);

                break;
        }
        GlStateManager.popMatrix();
    }

    /**
     * Combines the value and display title together
     *
     * @return the text that will be rendered
     */
    public String getDisplayString() {
        boolean showTitle = !getTitleText().get().trim().isEmpty();
        StringBuilder builder = new StringBuilder();
        if (getBrackets().get())
            builder.append("[");
        if (showTitle && !getInvertedTitle().get())
            builder.append(getTitleText().get()).append(": ");
        builder.append(getValue());
        if (showTitle && getInvertedTitle().get())
            builder.append(" ").append(getTitleText().get());
        if (getBrackets().get())
            builder.append("]");
        return builder.toString();
    }

    /**
     * @return the text to display on the HUD element
     */
    protected abstract String getValue();

    @Override
    protected float getHitBoxWidth() {
        return Math.max(mc.fontRendererObj.getStringWidth(getDisplayString()), 10);
    }

    @Override
    protected float getHitBoxHeight() {
        return mc.fontRendererObj.FONT_HEIGHT;
    }

    @Override
    protected void registerDefaultSettings() {
        super.registerDefaultSettings();

        addSettings(invertedTitle = new BooleanSetting("Invert Title", "Display", "If the title is rendered after the value.", false));
    }

    @Override
    public void loadJsonOld(BetterJsonObject root) {
        invertedTitle.set(root.optBoolean("inverted", false));

        super.loadJsonOld(root);
    }

}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/type/SimpleTextElement.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.type

import dev.isxander.evergreenhud.elements.RenderOrigin
import dev.isxander.evergreenhud.event.ClientTickEvent
import dev.isxander.settxi.impl.OptionContainer
import dev.isxander.evergreenhud.utils.drawString
import dev.isxander.evergreenhud.utils.mc
import dev.isxander.settxi.impl.option
import net.minecraft.client.util.math.MatrixStack
import kotlin.math.max

abstract class SimpleTextElement
@JvmOverloads constructor(title: String, cacheTime: Int = 5) : TextElement(title, cacheTime) {
    var titleLocation by option(TitleLocation.BEGINNING) {
        name = "Title Location"
        category = "Text"
        description = "Where to display the title."
    }

    val cachedDisplayString by eventReturnable<ClientTickEvent, String>("Calculating...", { clientTicks == 0 }) {
        displayString
    }

    protected abstract fun calculateValue(): String

    private val displayString: String
        get() {
            val showTitle = title.trim().isNotEmpty()
            var builder = ""

            if (brackets)
                builder += "["

            if (showTitle && titleLocation == TitleLocation.BEGINNING)
                builder += "${title}: "

            builder += calculateValue()

            if (showTitle && titleLocation == TitleLocation.END)
                builder += " $title"

            if (brackets)
                builder += "]"

            return builder
        }

    override val hitboxWidth: Float
        get() = max(mc.textRenderer.getWidth(cachedDisplayString), 10).toFloat()
    override val hitboxHeight: Float
        get() = max(mc.textRenderer.fontHeight, 10).toFloat()

    override fun render(matrices: MatrixStack, renderOrigin: RenderOrigin) {
        super.render(matrices, renderOrigin)

        matrices.push()
        matrices.scale(position.scale, position.scale, 1f)

        var x = position.rawX / position.scale
        val y = position.rawY / position.scale

        if (alignment == Alignment.RIGHT)
            x -= mc.textRenderer.getWidth(cachedDisplayString)

        drawString(
            matrices,
            cachedDisplayString!!,
            x, y,
            textColor.rgba,
            centered = alignment == Alignment.CENTER,
            shadow = textStyle == TextStyle.SHADOW,
            bordered = textStyle == TextStyle.BORDER,
            chroma = textColor.chroma.hasChroma, chromaSpeed = textColor.chroma.chromaSpeed
        )

        matrices.pop()
    }

    object TitleLocation : OptionContainer() {
        val BEGINNING = option("Beginning", "Show the title at the start of the text.")
        val END = option("End", "Show the title at the end of text.")
    }
}
